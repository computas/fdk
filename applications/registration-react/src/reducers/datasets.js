import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../constants/ActionTypes';

export default function datasets(
  state = { isFetchingDatasets: false, datasetItems: null },
  action
) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        isFetchingDatasets: true
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: action.response.data
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: null
      };
    default:
      return state;
  }
}
