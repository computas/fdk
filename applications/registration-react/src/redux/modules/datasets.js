import _ from 'lodash';
import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../../constants/ActionTypes';
import { fetchActions } from '../fetchActions';

function shouldFetch(metaState) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      (metaState.lastFetch || 0) < Date.now() - threshold)
  );
}

export function fetchDatasetsIfNeeded(catalogId) {
  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), ['datasets', catalogId, 'meta'])) &&
    dispatch(
      fetchActions(`/catalogs/${catalogId}/datasets?size=100`, [
        { type: DATASETS_REQUEST, meta: { catalogId } },
        { type: DATASETS_SUCCESS, meta: { catalogId } },
        { type: DATASETS_FAILURE, meta: { catalogId } }
      ])
    );
}

const initialState = {};

export default function datasets(state = initialState, action) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: true,
            lastFetch: null
          }
        }
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        [action.meta.catalogId]: {
          items: _.get(action.payload, ['_embedded', 'datasets']),
          meta: {
            isFetching: false,
            lastFetch: Date.now()
          }
        }
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: false,
            lastFetch: null
          }
        }
      };
    default:
      return state;
  }
}

export const getDatasetItemsByCatalogId = (datasets, catalogId) =>
  _.get(datasets, [catalogId, 'items']);
