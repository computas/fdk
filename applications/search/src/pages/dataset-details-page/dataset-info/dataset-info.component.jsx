import React from 'react';
import PropTypes from 'prop-types';
import Moment from 'react-moment';
import _ from 'lodash';

import localization from '../../../lib/localization';
import { getTranslateText } from '../../../lib/translateText';
import { ListRegular } from '../../../components/list-regular/list-regular.component';
import { TwoColRow } from '../../../components/list-regular/twoColRow/twoColRow';
import { LinkExternal } from '../../../components/link-external/link-external.component';
import './dataset-info.scss';
import { getReferenceDataByUri } from '../../../redux/modules/referenceData';
import { REFERENCEDATA_REFERENCETYPES } from '../../../constants/constants';

const renderReferences = (references, referencedItems, referenceData) => {
  const children = items =>
    items.map(item => {
      const referencedType = getReferenceDataByUri(
        referenceData,
        REFERENCEDATA_REFERENCETYPES,
        _.get(item, ['referenceType', 'uri'])
      );

      let referencedItem;
      if (referencedItems) {
        referencedItem = referencedItems.filter(referencedItem => {
          if (_.get(referencedItem, 'uri')) {
            return referencedItem.uri === _.get(item, ['source', 'uri']);
          }
          return null;
        });
      }

      return (
        <div
          key={`reference-${_.get(item, ['source', 'uri'])}`}
          className="d-flex list-regular--item"
        >
          <div className="col-4 pl-0 fdk-text-strong">
            {referencedType && referencedType.length > 0
              ? getTranslateText(_.get(referencedType[0], ['prefLabel']))
              : _.get(item, ['referenceType', 'uri'])}
          </div>
          <div className="col-8">
            <LinkExternal
              uri={item.uri}
              prefLabel={
                getTranslateText(_.get(referencedItem, [0, 'title'])) ||
                _.get(item, ['source', 'uri'])
              }
            />
          </div>
        </div>
      );
    });

  if (!references) {
    return null;
  }
  return (
    <ListRegular title={localization.dataset.related}>
      {children(references)}
    </ListRegular>
  );
};

const renderSpatial = spatial => {
  if (!spatial) {
    return null;
  }
  const spatialItems = spatial =>
    spatial.map((item, index) => (
      <span key={index}>
        {index > 0 ? ', ' : ''}
        {getTranslateText(_.get(item, 'prefLabel'))}
      </span>
    ));

  return (
    <div className="d-flex list-regular--item">
      <div className="col-4 pl-0 fdk-text-strong">
        {localization.dataset.spatial}
      </div>
      <div className="col-8">{spatialItems(spatial)}</div>
    </div>
  );
};

const renderTemporal = temporal => {
  const children = items =>
    items.map(item => (
      <React.Fragment key={`temporal-${_.get(item, 'startDate')}`}>
        {_.get(item, 'startDate') && (
          <TwoColRow
            col1={localization.dataset.periodFrom}
            col2={<Moment format="DD.MM.YYYY">{item.startDate}</Moment>}
          />
        )}
        {_.get(item, 'endDate') && (
          <TwoColRow
            col1={localization.dataset.periodTo}
            col2={<Moment format="DD.MM.YYYY">{item.endDate}</Moment>}
          />
        )}
      </React.Fragment>
    ));

  if (!temporal) {
    return null;
  }
  return children(temporal);
};

const renderRestrictions = (spatial, temporal) => {
  if (!(spatial || temporal)) {
    return null;
  }

  return (
    <ListRegular title={localization.dataset.restrictions}>
      {renderSpatial(spatial)}
      {renderTemporal(temporal)}
    </ListRegular>
  );
};

export const DatasetInfo = props => {
  const { datasetItem, referenceData, referencedItems } = props;
  const { references, spatial, temporal } = datasetItem || {};
  return (
    <React.Fragment>
      {renderReferences(references, referencedItems, referenceData)}
      {renderRestrictions(spatial, temporal)}
    </React.Fragment>
  );
};

DatasetInfo.defaultProps = {
  datasetItem: null,
  referenceData: null,
  referencedItems: null
};

DatasetInfo.propTypes = {
  datasetItem: PropTypes.object,
  referenceData: PropTypes.object,
  referencedItems: PropTypes.array
};
