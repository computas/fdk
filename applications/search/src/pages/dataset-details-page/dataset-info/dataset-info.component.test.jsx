import React from 'react';
import { shallow } from 'enzyme';
import { DatasetInfo } from './dataset-info.component';
import datasets from '../../../../test/fixtures/datasets';

test('should render DatasetInfo correctly with no props', () => {
  const minWrapper = shallow(<DatasetInfo />);
  expect(minWrapper).toMatchSnapshot();
});

test('should render DatasetInfo correctly', () => {
  const { _source } = datasets[0];
  const defaultProps = {
    issued: _source.issued,
    accrualPeriodicity: _source.accrualPeriodicity.prefLabel.nb,
    provenance: _source.provenance.prefLabel.nb,
    hasCurrentnessAnnotation: _source.hasCurrentnessAnnotation.hasBody.no,
    spatial: _source.spatial,
    temporal: _source.temporal,
    language: _source.language,
    references: _source.references
  };
  const wrapper = shallow(<DatasetInfo {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render DatasetInfo correctly', () => {
  const { _source } = datasets[1];
  const defaultProps = {
    issued: _source.issued,
    accrualPeriodicity: _source.accrualPeriodicity.prefLabel.nb,
    provenance: _source.provenance.prefLabel.nb,
    hasCurrentnessAnnotation: _source.hasCurrentnessAnnotation.hasBody.no,
    spatial: _source.spatial,
    temporal: _source.temporal,
    language: _source.language,
    references: _source.references
  };
  const wrapper = shallow(<DatasetInfo {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
