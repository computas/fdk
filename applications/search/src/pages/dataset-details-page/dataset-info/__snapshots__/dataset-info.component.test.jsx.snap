// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DatasetInfo correctly 1`] = `
<React.Fragment>
  <ListRegular
    bottomMargin={true}
    title="Related datasets"
  >
    <div
      className="d-flex list-regular--item"
      key="reference-http://referenced/dataset"
    >
      <div
        className="col-4 pl-0 fdk-text-strong"
      >
        http://www.w3.org/2002/07/hasVersion
      </div>
      <div
        className="col-8"
      >
        <LinkExternal
          prefLabel="http://referenced/dataset"
          uri=""
        />
      </div>
    </div>
  </ListRegular>
  <ListRegular
    bottomMargin={true}
    title="Restrictions"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-4 pl-0 fdk-text-strong"
      >
        Spatial
      </div>
      <div
        className="col-8"
      >
        <span
          key="0"
        >
          Asker
        </span>
        <span
          key="1"
        >
          , 
          Bærum
        </span>
        <span
          key="2"
        >
          , 
          Hurum
        </span>
        <span
          key="3"
        >
          , 
          Røyken
        </span>
      </div>
    </div>
    <React.Fragment
      key="temporal-1483228800000"
    >
      <TwoColRow
        col1="From"
        col2={
          <t
            ago={false}
            calendar={false}
            decimal={false}
            element="time"
            filter={[Function]}
            format="DD.MM.YYYY"
            fromNow={false}
            interval={60000}
            onChange={[Function]}
            toNow={false}
            unit={null}
            unix={false}
            utc={false}
          >
            1483228800000
          </t>
        }
      />
      <TwoColRow
        col1="To"
        col2={
          <t
            ago={false}
            calendar={false}
            decimal={false}
            element="time"
            filter={[Function]}
            format="DD.MM.YYYY"
            fromNow={false}
            interval={60000}
            onChange={[Function]}
            toNow={false}
            unit={null}
            unix={false}
            utc={false}
          >
            1514764799000
          </t>
        }
      />
    </React.Fragment>
  </ListRegular>
</React.Fragment>
`;

exports[`should render DatasetInfo correctly with no props 1`] = `<React.Fragment />`;
